### PR TITLE
Make the "random snail fact" a server-side operation

### PR DESCRIPTION
**Pull request by ananyeet**
Add a function and change the "Random Snail Fact" button to perform the choosing of a random snail fact on the server side instead of on the client side. 

No visual changes. Tested by repeated usage of the button, as well as console logging to make sure the correct function in script.js is used.

### CODE REVIEW

File: portfolio/src/main/webapp/script.js [hunk 1]
```diff
@@ -39,3 +39,10 @@ function randomSnailFact() {
     const container = document.getElementById('snail-fact-container');
     container.innerText = chosen_fact;
 }
+
+async function randomSnailFact_server() {
+    console.log("requesting a snail fact...");
+    const response = await fetch('/data');
+    const fact = await response.text();
+    document.getElementById('snail-fact-container').innerText = fact;
```
Comment (beiebi0) [inline portfolio/src/main/webapp/script.js:47]: what's the expected behavior here if the server call fails?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -23,10 +25,29 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+    private List<String> snailFacts;
```
[the remark this answers is not in the document]
Comment (ananyeet) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:28]: Which part? the variable name is written in camelcase as defined [here](https://google.github.io/styleguide/javaguide.html#s5.2.7-local-variable-names) and I followed the [example](https://github.com/ananyeet/step/blob/master/walkthroughs/week-3-server/examples/random-quotes/src/main/java/com/google/sps/servlets/RandomQuoteServlet.java) for everything else.

[the remark this answers is not in the document]
Comment (beiebi0) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:28]: yeah I should be more specific. The indentation should be 2 spaces. I believe you have 4 spaces for the entire file.

File: portfolio/src/main/webapp/script.js (hunk 1 above)
Comment (ananyeet) [reply portfolio/src/main/webapp/script.js:47]: I think the default fail response is "ok"..? ([Source](https://www.tjvantoll.com/2015/09/13/fetch-and-errors/) ) If so, for now that's a perfectly acceptable response.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -23,10 +25,30 @@
...
+        "knives, and rough surfaces without being scratched.");
+    snailFacts.add("When there isn't enough moisture in the area, snails " +
+        "go into a deep sleep that can last as long as 3 years.");
+    snailFacts.add("In the winter, snails go into hibernation.");
+  }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Serve up them snail facts. 
+
```
Comment (beiebi0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:47]: remove the empty line here.

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -14,14 +14,21 @@
...
- 
-    const chosen_fact = facts[Math.floor(Math.random() * facts.length)];
- 
-    const container = document.getElementById('snail-fact-container');
-    container.innerText = chosen_fact;
+  const facts = ["To adapt to city life, snails in urban centres are more likely to be yellow than pink.", 
+  "Snails can't move backwards.", 
+  "Snails can slide over sharp edges such as blades, knives, and rough surfaces without being scratched.",
+  "When there isn't enough moisture in the area, snails go into a deep sleep that can last as long as 3 years.", 
+  "In the winter, snails go into hibernation."]
```
Comment (beiebi0) [inline portfolio/src/main/webapp/script.js:21]: missing ";"

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -14,14 +14,21 @@
...
+
+  const chosen_fact = facts[Math.floor(Math.random() * facts.length)];
```
Comment (beiebi0) [inline portfolio/src/main/webapp/script.js:23]: chosenFact

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (beiebi0) [inline portfolio/src/main/webapp/script.js:21]:
```suggestion
    const facts = [
      'To adapt to city life, snails in urban centres are more likely to be yellow than pink.',
      'Snails can\'t move backwards.',
      'Snails can slide over sharp edges such as blades, knives, and rough surfaces without being scratched.',
      'When there isn\'t enough moisture in the area, snails go into a deep sleep that can last as long as 3 years.',
      'In the winter, snails go into hibernation.'
    ];
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -23,10 +25,30 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  private List<String> snailFacts;
```
Comment (beiebi0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:28]: This is fine. If this facts won't be changed after init(). Its better to have 
```suggestion
  private static final List<String> snailFacts = ImmutableList.of(factString1,factString2, ...);
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -23,10 +25,30 @@
...
+    snailFacts.add("Snails can slide over sharp edges such as blades, " +
+        "knives, and rough surfaces without being scratched.");
+    snailFacts.add("When there isn't enough moisture in the area, snails " +
+        "go into a deep sleep that can last as long as 3 years.");
+    snailFacts.add("In the winter, snails go into hibernation.");
+  }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Serve up them snail facts.
```
Comment (beiebi0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:46]: you mean "the"? or could you please make it more clear who are "them" here?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -23,10 +26,25 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  private static final List<String> snailFacts = ImmutableList.of(
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:29]: Constant names should be in SNAKE_UPPER_CASE format. https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -23,10 +26,25 @@
...
+    "Snails can slide over sharp edges such as blades, " +
+      "knives, and rough surfaces without being scratched.",
+    "When there isn't enough moisture in the area, snails " +
+      "go into a deep sleep that can last as long as 3 years.", 
+    "In the winter, snails go into hibernation."
+  );
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Serve ups snail facts.
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:43]: Remove "s" from "ups"?

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -14,14 +14,23 @@
...
+      'Snails can\'t move backwards.',
+      'Snails can slide over sharp edges such as blades, knives, and rough surfaces without being scratched.',
+      'When there isn\'t enough moisture in the area, snails go into a deep sleep that can last as long as 3 years.',
+      'In the winter, snails go into hibernation.'
+    ];
+
+  const chosenFact = facts[Math.floor(Math.random() * facts.length)];
+
+  const container = document.getElementById('snail-fact-container');
+  container.innerText = chosenFact;
```
Comment (Hiram-Silvey) [inline portfolio/src/main/webapp/script.js:28]: Shouldn't this be removed if it's being replaced by the server side snail facts function?

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -14,14 +14,23 @@
...
+}
+
+async function getRandomSnailFactFromServer() {
+  console.log("requesting a snail fact...");
```
Comment (Hiram-Silvey) [inline portfolio/src/main/webapp/script.js:32]: Remove extraneous logs from client-side code after testing is complete.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -14,6 +14,9 @@
 
 package com.google.sps.servlets;
 
+import com.google.common.collect.ImmutableList; 
+import java.util.ArrayList;
```
Comment (Hiram-Silvey) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:18]: Is this import used? If not, please remove it.

File: portfolio/src/main/webapp/script.js (hunk 1 above)
Comment (beiebi0) [reply portfolio/src/main/webapp/script.js]: sg

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 12 above)
Comment (beiebi0) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:18]: +1

Comment (ananyeet) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: it is used! for the static final list of snail facts :)

**Pull request merged**